feat(trace-view): Link docs for missing services

wmak opened the pull request:
- This adds a new `(???)` node to quick trace when we think that there's
  probably a service that isn't being instrumented by Sentry
- This only shows up when the trace has exactly one event, and there's
  an HTTP span without a domain, or in the same domain as the current
  transaction
- The dropdown has a dismiss that is permanent, we can probably change
  the key if we decide we want to start showing this again after EA or
  when we add more methods for determining missing services
- [ ] TODO: update reload

  
Preview:
<img width="258" alt="image" src="https://user-images.githubusercontent.com/4205004/113958471-280f4780-97ef-11eb-8696-b49f968a8746.png">
<img width="253" alt="image" src="https://user-images.githubusercontent.com/4205004/113958485-2fceec00-97ef-11eb-9657-305ed0e8768e.png">

wmak committed:
feat(trace-view): Link docs for missing services

- This adds a new (???) node to quick trace when we think that there's
  probably a service that isn't being instrumented by Sentry
- This only shows up when the trace has exactly one event, and there's
  an HTTP span without a domain, or in the same domain as the current
  transaction
- The dropdown has a dismiss that is permanent, we can probably change
  the key if we decide we want to start showing this again after EA or
  when we add more methods for determining missing services

benvinegar commented:
I'm not sure we should have a dismiss button until people actually tell us it is irritating. I'm not sure it is TBH.

benvinegar commented:
Ah, I see it's only for the current user. Not a big deal then.

wmak commented on src/sentry/api/endpoints/organization_events_trace.py:
```diff
@@ -81,6 +82,18 @@ def construct_span_map(self, events, key):
+        child_domain = child["description"].split()[-1]
+        # string contains is significantly faster than urlparse
+        if (
+            domain in child_domain and urlparse(child_domain).netloc == domain
```
`urlparse('https://sentry.io/organizations/test/performance/').netloc`
100000 loops, best of 3: 2.74 usec per loop
vs
`'sentry.io' in 'https://sentry.io/organizations/test/performance/'`
10000000 loops, best of 3: 0.058 usec per loop

dashed commented on src/sentry/static/sentry/app/components/quickTrace/index.tsx:
```diff
@@ -446,3 +452,62 @@ function StyledEventNode({text, hoverText, to, onClick, type = 'white'}: EventNo
+          anchorRight={anchor === 'right'}
+        >
+          <DropdownItem first width="small">
+            <ExternalDropdownLink href="https://docs.sentry.io/platforms/javascript/performance/connect-services/">
```
Will this only be available for JavaScript projects?

wmak replied:
Hmm good point, this can appear for any project, I'll backlog something to add a product level doc since connecting service isn't JavaScript specific :+1:

dashed commented on src/sentry/static/sentry/app/components/quickTrace/index.tsx:
```diff
@@ -446,3 +452,62 @@ function StyledEventNode({text, hoverText, to, onClick, type = 'white'}: EventNo
+        >
+          <DropdownItem first width="small">
+            <ExternalDropdownLink href="https://docs.sentry.io/platforms/javascript/performance/connect-services/">
+              Connect to a service
```
```suggestion
              t('Connect to a service')
```

dashed commented on src/sentry/static/sentry/app/components/quickTrace/index.tsx:
```diff
@@ -446,3 +452,62 @@ function StyledEventNode({text, hoverText, to, onClick, type = 'white'}: EventNo
+            </ExternalDropdownLink>
+          </DropdownItem>
+          <DropdownItem onSelect={this.dismissMissingService} width="small">
+            Dismiss
```
```suggestion
            {t('Dismiss')}
```

wmak committed:
ref: Apply suggestions from code review

Co-authored-by: Alberto Leal <mail4alberto@gmail.com>

wmak committed:
ref: Include localhost as a potential missing service

- Since a request to something in localhost probably means you still own
  the service too

Zylphrex commented on src/sentry/api/endpoints/organization_events_trace.py:
```diff
@@ -157,7 +171,15 @@ def get(self, request, organization, trace_id):
+                warning_extra,
+                event_id,
+                detailed=detailed,
+                check_missing=len(result["data"]) == 1,
```
Let's be explicit about what this is going to be checking for.

```suggestion
                check_missing_backend=len(result["data"]) == 1,
```

wmak replied:
My plan is to eventually re-use this for other missing checks, which means that this won't always be backend specific. Can rename to `check_missing_service` if the concern is with the name not being explicit enough.

Zylphrex commented on src/sentry/static/sentry/app/components/quickTrace/index.tsx:
```diff
@@ -446,3 +452,62 @@ function StyledEventNode({text, hoverText, to, onClick, type = 'white'}: EventNo
+    const {hideMissing} = this.state;
+    const {anchor} = this.props;
+    if (hideMissing) {
+      return <React.Fragment />;
```
```suggestion
      return null;
```

wmak committed:
Update src/sentry/static/sentry/app/components/quickTrace/index.tsx

Co-authored-by: Tony <txiao@sentry.io>

getsantry[bot] committed:
style(lint): Auto commit lint changes

wmak commented:
Closing for now, going to change this to show the `???` based on txn count == 1, and project type.
This is because I assumed that we made spans for all http requests, but we only do this for the domains in `TracingOrigins`.
see [docs](https://docs.sentry.io/platforms/javascript/guides/react/performance/included-instrumentation/#shouldcreatespanforrequest)
This meant that if the http span showed up, chances are the user already had instrumentation correctly setup
